File: backend/app/api/v1/billboards.py
"""Billboard endpoints"""

from typing import Optional
from fastapi import APIRouter, Depends, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from uuid import UUID
from app.core.database import get_db
from app.core.auth import get_current_user, get_current_owner, get_optional_user
from app.schemas.billboard import (
    BillboardCreate,
    BillboardUpdate,
    BillboardResponse,
    BillboardDetailResponse,
    BillboardListItem,
    BillboardFilters
)
from app.schemas.payment import BillboardListingStatus
from app.schemas.common import ResponseModel, PaginationParams, PaginatedResponse
from app.schemas.user import UserPublicProfile
from app.services.billboard_service import BillboardService
from app.services.payment_service import payment_service
from app.models.user import User
from app.core.exceptions import ForbiddenException


router = APIRouter()


@router.post(
    "",
    response_model=ResponseModel[BillboardResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Create billboard",
    description="Create a new billboard listing (owners only)"
)
async def create_billboard(
    billboard_data: BillboardCreate,
    current_user: User = Depends(get_current_owner),
    db: AsyncSession = Depends(get_db)
):
    """Create a new billboard listing"""
    
    billboard = await BillboardService.create_billboard(
        db,
        current_user,
        billboard_data
    )
    
    return ResponseModel(
        success=True,
        data=BillboardResponse.model_validate(billboard),
        message="Billboard created successfully"
    )


@router.get(
    "",
    response_model=ResponseModel[PaginatedResponse[BillboardListItem]],
    summary="Browse billboards",
    description="Get billboards with filters and pagination"
)
async def browse_billboards(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    location: Optional[str] = None,
    type: Optional[str] = None,
    availability: Optional[str] = None,
    min_weekly_rate: Optional[int] = None,
    max_weekly_rate: Optional[int] = None,
    min_monthly_rate: Optional[int] = None,
    max_monthly_rate: Optional[int] = None,
    min_views: Optional[int] = None,
    is_promoted: Optional[bool] = None,
    search: Optional[str] = None,
    owner_id: Optional[UUID] = None,
    # Proximity search parameters
    near_lat: Optional[float] = Query(None, description="Latitude for proximity search"),
    near_lng: Optional[float] = Query(None, description="Longitude for proximity search"),
    radius_km: Optional[float] = Query(5.0, ge=0.1, le=100, description="Search radius in km"),
    current_user: Optional[User] = Depends(get_optional_user),
    db: AsyncSession = Depends(get_db)
):
    """Browse billboards with advanced filtering"""
    
    # Convert type string to BillboardType enum if provided
    billboard_type = None
    if type:
        try:
            from app.models.billboard import BillboardType
            billboard_type = BillboardType(type)
        except ValueError:
            pass
    
    # Convert availability string to boolean
    is_available = None
    if availability == "available":
        is_available = True
    elif availability == "unavailable":
        is_available = False
    
    filters = BillboardFilters(
        location=location,
        billboard_type=billboard_type,
        is_available=is_available,
        min_weekly_rate=min_weekly_rate,
        max_weekly_rate=max_weekly_rate,
        min_monthly_rate=min_monthly_rate,
        max_monthly_rate=max_monthly_rate,
        min_views=min_views,
        search=search,
        owner_id=owner_id,
        near_lat=near_lat,
        near_lng=near_lng,
        radius_km=radius_km if near_lat and near_lng else None
    )
    
    pagination = PaginationParams(page=page, limit=limit)
    
    billboards, total = await BillboardService.get_billboards(
        db,
        filters,
        pagination,
        current_user
    )
    
    items = [BillboardListItem.model_validate(b) for b in billboards]
    
    paginated = PaginatedResponse.create(
        items=items,
        total=total,
        page=page,
        limit=limit
    )
    
    return ResponseModel(
        success=True,
        data=paginated,
        message=f"Found {total} billboards"
    )


@router.get(
    "/{billboard_id}",
    response_model=ResponseModel[BillboardDetailResponse],
    summary="Get billboard details",
    description="Get detailed information about a specific billboard"
)
async def get_billboard(
    billboard_id: UUID,
    current_user: Optional[User] = Depends(get_optional_user),
    db: AsyncSession = Depends(get_db)
):
    """Get billboard details"""
    
    billboard = await BillboardService.get_billboard_by_id(
        db,
        billboard_id,
        load_owner=True
    )
    
    # Check if current user has saved this billboard
    is_saved = False
    if current_user:
        is_saved = await BillboardService.is_billboard_saved(
            db,
            current_user,
            billboard_id
        )
    
    # Build response
    response_data = BillboardDetailResponse.model_validate(billboard)
    response_data.is_saved = is_saved
    response_data.total_saves = len(billboard.saved_by) if billboard.saved_by else 0
    
    return ResponseModel(
        success=True,
        data=response_data,
        message="Billboard retrieved successfully"
    )


@router.patch(
    "/{billboard_id}",
    response_model=ResponseModel[BillboardResponse],
    summary="Update billboard",
    description="Update billboard information (owner only)"
)
async def update_billboard(
    billboard_id: UUID,
    update_data: BillboardUpdate,
    current_user: User = Depends(get_current_owner),
    db: AsyncSession = Depends(get_db)
):
    """Update billboard"""
    
    billboard = await BillboardService.get_billboard_by_id(db, billboard_id)
    
    updated_billboard = await BillboardService.update_billboard(
        db,
        billboard,
        current_user,
        update_data
    )
    
    return ResponseModel(
        success=True,
        data=BillboardResponse.model_validate(updated_billboard),
        message="Billboard updated successfully"
    )


@router.delete(
    "/{billboard_id}",
    response_model=ResponseModel[dict],
    summary="Delete billboard",
    description="Delete billboard (owner only, soft delete)"
)
async def delete_billboard(
    billboard_id: UUID,
    current_user: User = Depends(get_current_owner),
    db: AsyncSession = Depends(get_db)
):
    """Delete billboard"""
    
    billboard = await BillboardService.get_billboard_by_id(db, billboard_id)
    
    await BillboardService.delete_billboard(db, billboard, current_user)
    
    return ResponseModel(
        success=True,
        data={"message": "Billboard deleted successfully"},
        message="Billboard deleted"
    )


@router.post(
    "/{billboard_id}/view",
    response_model=ResponseModel[dict],
    summary="Increment view count",
    description="Increment billboard view count"
)
async def increment_billboard_views(
    billboard_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """Increment billboard view count"""
    
    await BillboardService.increment_views(db, billboard_id)
    
    return ResponseModel(
        success=True,
        data={"message": "View count incremented"},
        message="View recorded"
    )


@router.post(
    "/{billboard_id}/save",
    response_model=ResponseModel[dict],
    summary="Save billboard",
    description="Save/bookmark a billboard"
)
async def save_billboard(
    billboard_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Save/bookmark a billboard"""
    
    await BillboardService.save_billboard(db, current_user, billboard_id)
    
    return ResponseModel(
        success=True,
        data={"message": "Billboard saved successfully"},
        message="Billboard saved"
    )


@router.delete(
    "/{billboard_id}/save",
    response_model=ResponseModel[dict],
    summary="Unsave billboard",
    description="Remove billboard from saved/bookmarks"
)
async def unsave_billboard(
    billboard_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Unsave/unbookmark a billboard"""
    
    await BillboardService.unsave_billboard(db, current_user, billboard_id)
    
    return ResponseModel(
        success=True,
        data={"message": "Billboard removed from saved"},
        message="Billboard unsaved"
    )


@router.get(
    "/saved/list",
    response_model=ResponseModel[PaginatedResponse[BillboardListItem]],
    summary="Get saved billboards",
    description="Get user's saved/bookmarked billboards"
)
async def get_saved_billboards(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Get saved billboards"""
    
    pagination = PaginationParams(page=page, limit=limit)
    
    billboards, total = await BillboardService.get_saved_billboards(
        db,
        current_user,
        pagination
    )
    
    items = [BillboardListItem.model_validate(b) for b in billboards]
    
    paginated = PaginatedResponse.create(
        items=items,
        total=total,
        page=page,
        limit=limit
    )
    
    return ResponseModel(
        success=True,
        data=paginated,
        message=f"Found {total} saved billboards"
    )


@router.get(
    "/owner/my-billboards",
    response_model=ResponseModel[PaginatedResponse[BillboardResponse]],
    summary="Get my billboards",
    description="Get all billboards owned by current user"
)
async def get_my_billboards(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    current_user: User = Depends(get_current_owner),
    db: AsyncSession = Depends(get_db)
):
    """Get current owner's billboards"""
    
    pagination = PaginationParams(page=page, limit=limit)
    
    billboards, total = await BillboardService.get_owner_billboards(
        db,
        current_user.id,
        pagination
    )
    
    items = [BillboardResponse.model_validate(b) for b in billboards]
    
    paginated = PaginatedResponse.create(
        items=items,
        total=total,
        page=page,
        limit=limit
    )
    
    return ResponseModel(
        success=True,
        data=paginated,
        message=f"Found {total} billboards"
    )


@router.get(
    "/{billboard_id}/listing-status",
    response_model=ResponseModel[BillboardListingStatus],
    summary="Get billboard listing status",
    description="Get the listing access status for a billboard (owner only)"
)
async def get_billboard_listing_status(
    billboard_id: UUID,
    current_user: User = Depends(get_current_owner),
    db: AsyncSession = Depends(get_db)
):
    """Get billboard listing access status"""
    
    # Verify ownership
    billboard = await BillboardService.get_billboard_by_id(db, billboard_id)
    
    if billboard.owner_id != current_user.id:
        raise ForbiddenException(detail="You don't have permission to view this billboard's listing status")
    
    # Get listing status from payment service
    listing_status = await payment_service.get_billboard_listing_status(
        db,
        billboard_id
    )
    
    return ResponseModel(
        success=True,
        data=listing_status,
        message="Listing status retrieved successfully"
    )

File: client/app/api/endpoints/billboards.ts
/**
 * Billboard API Endpoints
 * Billboard browsing, CRUD, and interaction endpoints
 */

import apiClient from '../index';
import type {
  ApiResponse,
  PaginatedResponse,
  Billboard,
  BillboardFilters,
  BillboardCreateInput,
  BillboardUpdateInput,
  BillboardListingStatus,
} from '@/lib/types';

// Transform functions for snake_case <-> camelCase
const toCamelCase = (data: any): Billboard => ({
  id: data.id,
  title: data.title,
  description: data.description,
  location: data.location,
  fullAddress: data.full_address,
  coordinates: data.coordinates,
  billboardType: data.billboard_type,
  widthFt: data.width_ft,
  heightFt: data.height_ft,
  orientation: data.orientation,
  illumination: data.illumination,
  images: data.images || [],
  isAvailable: data.is_available,
  isActive: data.is_active,
  views: data.views,
  weeklyRate: data.weekly_rate,
  monthlyRate: data.monthly_rate,
  printingFee: data.printing_fee,
  flightFee: data.flight_fee,
  minimumDuration: data.minimum_duration,
  features: data.features || [],
  nearbyLandmarks: data.nearby_landmarks || [],
  availableFrom: data.available_from,
  availableTo: data.available_to,
  ownerId: data.owner_id,
  owner: data.owner ? {
    id: data.owner.id,
    fullName: data.owner.full_name,
    phoneNumber: data.owner.phone_number,
    companyName: data.owner.company_name,
    bio: data.owner.bio,
    profileImage: data.owner.profile_image,
    userType: data.owner.user_type,
  } : undefined,
  isSaved: data.is_saved,
  totalSaves: data.total_saves,
  contactName: data.contact_name,
  contactPhone: data.contact_phone,
  createdAt: data.created_at,
  updatedAt: data.updated_at,
});

const toSnakeCase = (data: BillboardCreateInput | BillboardUpdateInput) => {
  return {
    title: data.title,
    description: data.description,
    location: data.location,
    full_address: data.fullAddress,
    coordinates: data.coordinates,
    billboard_type: data.billboardType,
    width_ft: data.widthFt,
    height_ft: data.heightFt,
    orientation: data.orientation,
    illumination: data.illumination,
    weekly_rate: data.weeklyRate,
    monthly_rate: data.monthlyRate,
    printing_fee: data.printingFee,
    flight_fee: data.flightFee,
    minimum_duration: data.minimumDuration,
    features: data.features || [],
    nearby_landmarks: data.nearbyLandmarks || [],
    available_from: data.availableFrom,
    available_to: data.availableTo,
    images: data.images || [],
    contact_name: data.contactName,
    contact_phone: data.contactPhone,
  };
};

const transformFilters = (filters: BillboardFilters) => ({
  location: filters.location,
  type: filters.billboardType,
  availability: filters.isAvailable !== undefined ? (filters.isAvailable ? 'available' : 'unavailable') : undefined,
  min_weekly_rate: filters.minWeeklyRate,
  max_weekly_rate: filters.maxWeeklyRate,
  min_monthly_rate: filters.minMonthlyRate,
  max_monthly_rate: filters.maxMonthlyRate,
  min_views: filters.minViews,
  search: filters.search,
  owner_id: filters.ownerId,
  // Proximity search parameters
  near_lat: filters.nearLat,
  near_lng: filters.nearLng,
  radius_km: filters.radiusKm,
});

export const billboardEndpoints = {
  /**
   * Browse billboards with filters and pagination
   */
  browse: async (
    filters: BillboardFilters,
    page = 1,
    limit = 20
  ): Promise<ApiResponse<PaginatedResponse<Billboard>>> => {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
      ...Object.fromEntries(
        Object.entries(transformFilters(filters))
          .filter(([_, v]) => v !== undefined && v !== '')
          .map(([k, v]) => [k, String(v)])
      ),
    });

    const response = await apiClient.get(`/billboards?${params}`);
    return {
      ...response.data,
      data: {
        ...response.data.data,
        items: response.data.data.items.map(toCamelCase),
      },
    };
  },

  /**
   * Get billboard by ID
   */
  getById: async (id: string): Promise<ApiResponse<Billboard>> => {
    const response = await apiClient.get(`/billboards/${id}`);
    return {
      ...response.data,
      data: toCamelCase(response.data.data),
    };
  },

  /**
   * Create a new billboard (owner only)
   */
  create: async (data: BillboardCreateInput): Promise<ApiResponse<Billboard>> => {
    const response = await apiClient.post('/billboards', toSnakeCase(data));
    return {
      ...response.data,
      data: toCamelCase(response.data.data),
    };
  },

  /**
   * Create a billboard on behalf of an owner (admin only)
   */
  adminCreate: async (data: BillboardCreateInput): Promise<ApiResponse<Billboard>> => {
    const response = await apiClient.post('/admin/billboards', toSnakeCase(data));
    return {
      ...response.data,
      data: toCamelCase(response.data.data),
    };
  },

  /**
   * Update billboard (owner only)
   */
  update: async (
    id: string,
    data: BillboardUpdateInput
  ): Promise<ApiResponse<Billboard>> => {
    const response = await apiClient.patch(`/billboards/${id}`, toSnakeCase(data));
    return {
      ...response.data,
      data: toCamelCase(response.data.data),
    };
  },

  /**
   * Delete billboard (owner only, soft delete)
   */
  delete: async (id: string): Promise<ApiResponse<{ message: string }>> => {
    const response = await apiClient.delete(`/billboards/${id}`);
    return response.data;
  },

  /**
   * Increment view count
   */
  incrementViews: async (id: string): Promise<ApiResponse<{ message: string }>> => {
    const response = await apiClient.post(`/billboards/${id}/view`);
    return response.data;
  },

  /**
   * Save/bookmark a billboard
   */
  save: async (id: string): Promise<ApiResponse<{ message: string }>> => {
    const response = await apiClient.post(`/billboards/${id}/save`);
    return response.data;
  },

  /**
   * Unsave/unbookmark a billboard
   */
  unsave: async (id: string): Promise<ApiResponse<{ message: string }>> => {
    const response = await apiClient.delete(`/billboards/${id}/save`);
    return response.data;
  },

  /**
   * Get saved billboards
   */
  getSaved: async (page = 1, limit = 20): Promise<ApiResponse<PaginatedResponse<Billboard>>> => {
    const response = await apiClient.get(`/billboards/saved/list?page=${page}&limit=${limit}`);
    return {
      ...response.data,
      data: {
        ...response.data.data,
        items: response.data.data.items.map(toCamelCase),
      },
    };
  },

  /**
   * Get my billboards (owner only)
   */
  getMyBillboards: async (
    page = 1,
    limit = 20
  ): Promise<ApiResponse<PaginatedResponse<Billboard>>> => {
    const response = await apiClient.get(`/billboards/owner/my-billboards?page=${page}&limit=${limit}`);
    return {
      ...response.data,
      data: {
        ...response.data.data,
        items: response.data.data.items.map(toCamelCase),
      },
    };
  },

  /**
   * Get listing status for a billboard (owner only)
   */
  getListingStatus: async (id: string): Promise<ApiResponse<BillboardListingStatus>> => {
    const response = await apiClient.get(`/billboards/${id}/listing-status`);
    const data = response.data;
    
    // Transform snake_case to camelCase
    if (data.success && data.data) {
      return {
        success: data.success,
        data: {
          billboardId: data.data.billboard_id,
          hasActiveListing: data.data.has_active_listing,
          accessStartsAt: data.data.access_starts_at,
          accessExpiresAt: data.data.access_expires_at,
          daysRemaining: data.data.days_remaining,
          isExpired: data.data.is_expired,
          isInGracePeriod: data.data.is_in_grace_period,
          gracePeriodExpiresAt: data.data.grace_period_expires_at,
          canRenew: data.data.can_renew,
          currentTier: data.data.current_tier ? {
            tierId: data.data.current_tier.tier_id,
            durationDays: data.data.current_tier.duration_days,
            priceGhs: data.data.current_tier.price_ghs,
            accessStartsAt: data.data.current_tier.access_starts_at,
            accessExpiresAt: data.data.current_tier.access_expires_at,
            isActive: data.data.current_tier.is_active,
          } : null,
        },
        message: data.message,
      };
    }
    
    return data;
  },
};
